Major fix for get_s for BlocksOnCylindrical along with with minor changes

VietAnhDao opened the pull request.

VietAnhDao committed:
Major fix for get_s for BlocksOnCylindrical along with with minor changes

VietAnhDao commented:
@KrisThielemans @danieldeidda @emikhay

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -95,21 +102,22 @@ SSRB(const ProjDataInfo& in_proj_data_info,
+  //     const float offset = in_proj_data_info_sptr->get_azimuthal_angle_offset() +
+  //       in_proj_data_info_sptr->get_azimuthal_angle_sampling() * (num_views_to_combine-1)/2.F;
+  //     out_proj_data_info_sptr->set_azimuthal_angle_offset(offset);
+  //   }
```
why commented out? shouldn't be the case

KrisThielemans commented on src/buildblock/GeometryBlocksOnCylindrical.cxx:
```diff
@@ -108,6 +108,7 @@ find_detection_position_given_cartesian_coordinate(DetectionPosition<>& det_pos,
    is not precisely pointing to the center of the crystal and
    then the det_pos cannot be found using the map
   */
+  // NEED RE-WORKING	
```
remove comment?

KrisThielemans commented on src/buildblock/GeometryBlocksOnCylindrical.cxx:
```diff
@@ -156,85 +166,84 @@ build_crystal_maps()
 	}
-
+	
+	/*
```
I'd rather remove all this code, as opposed to commenting it out

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -95,21 +102,22 @@ SSRB(const ProjDataInfo& in_proj_data_info,
+     ? in_proj_data_info.get_max_segment_num()
+     : max_in_segment_num_to_process
+     )-(num_segments_to_combine/2))/num_segments_to_combine;
+  //const int out_max_segment_num = 0;
```
remove

KrisThielemans commented on src/include/stir/ProjDataInfoBlocksOnCylindrical.inl:
```diff
@@ -79,7 +79,6 @@ ProjDataInfoBlocksOnCylindrical::get_m(const Bin& bin) const
 	LORInAxialAndNoArcCorrSinogramCoordinates<float> lor;
 	get_LOR(lor, bin);
 	//Parisa to Check
-	//std::cout<<"seg_ax_m_z1 = "<<bin.segment_num()<<"\t"<<bin.axial_pos_num()<<"\t"<<lor.z1()<<"\t"<<(lor.z1() + lor.z2())/2.<<std::endl;
```
remove

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -56,273 +55,268 @@
-                    %this->recon_template_par_filename);
-            return true;
-        }
+      warning(boost::format("ScatterEstimation: Error parsing reconstruction parameters file %1%. Aborting.") % this->recon_template_par_filename);
```
no need for reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -56,273 +55,268 @@
-                    %this->masking_parameters.filter_filename);
-            return true;
-        }
+      warning(boost::format("ScatterEstimation: Error parsing post filter parameters file %1%. Aborting.") % this->masking_parameters.filter_filename);
```
no need for reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -338,35 +332,33 @@ shared_ptr<const ProjData> ScatterEstimation::get_input_data() const
+void ScatterEstimation::set_output_scatter_estimate_prefix(const std::string &arg)
 {
   this->output_scatter_estimate_prefix = arg;
+  std::cout << arg;
```
remove

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -380,1030 +372,1052 @@ bool ScatterEstimation::already_setup() const
+    info("ScatterEstimation: Running SSRB on input data...");
+    shared_ptr<ProjDataInfo> proj_data_info_2d_sptr(
+        SSRB(*this->input_projdata_sptr->get_proj_data_info_sptr(),
+             1, 1, false, 0));
```
this will break stuff

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -380,1030 +372,1052 @@ bool ScatterEstimation::already_setup() const
+      this->apply_mask_in_place(*mask_image_ptr, this->masking_parameters);
+      this->mask_image_sptr.reset(mask_image_ptr);
+      if (this->mask_image_filename.size() > 0)
+        OutputFileFormat<DiscretisedDensity<3, float>>::default_sptr()->write_to_file(this->mask_image_filename, *this->mask_image_sptr);
```
no need for reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -380,1030 +372,1052 @@ bool ScatterEstimation::already_setup() const
-        std::string filename = extras_path.get_path() + "recon_0.hv";
-        current_activity_image_sptr = read_from_file<DiscretisedDensity<3,float> >(filename);
+      std::string out_filename = extras_path.get_path() + "initial_activity_image";
+      OutputFileFormat<DiscretisedDensity<3, float>>::default_sptr()->write_to_file(out_filename, *this->current_activity_image_sptr);
```
no need for reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterEstimation.cxx:
```diff
@@ -380,1030 +372,1052 @@ bool ScatterEstimation::already_setup() const
+                                      *normalisation_factors_sptr,
+                                      *this->mask_projdata_sptr, local_min_scale_value,
+                                      local_max_scale_value, this->half_filter_width,
+                                      spline_type, false);
```
why?

VietAnhDao replied:
Interleaving made upscaling single scatter simulation worse, turn off interleaving to alleviate the problem.

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -215,14 +219,15 @@ void
 ScatterSimulation::set_defaults()
 {
     this->attenuation_threshold =  0.01f ;
+    // CHANGED THIS TO FALSE
```
remove, as you didn't change it

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -215,14 +219,15 @@ void
     this->zoom_size_z = -1;
     this->downsample_scanner_bool = false;
-    this->downsample_scanner_dets = 64;
+    this->downsample_scanner_dets = -1;
```
will break things. we should keep the old default I think. Can't this be set via the .par file?

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -780,25 +803,26 @@ set_exam_info_sptr(const shared_ptr<const ExamInfo> arg)
     if (new_num_rings <= 0)
     {
-	if(downsample_scanner_rings > 0)
+	if(downsample_scanner_rings > 0){
             new_num_rings = downsample_scanner_rings;
+        }
```
avoid reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -780,25 +803,26 @@ set_exam_info_sptr(const shared_ptr<const ExamInfo> arg)
     if (new_num_dets <= 0)
     {
-        if(downsample_scanner_dets > 0)
+        if(downsample_scanner_dets > 0){
             new_num_dets = downsample_scanner_dets;
+        }
```
avoid reformatting

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -891,7 +913,7 @@ void
 set_randomly_place_scatter_points(const bool arg)
 {
-    randomly_place_scatter_points = arg;
+    randomly_place_scatter_points = false;
```
remove

KrisThielemans commented on src/scatter_buildblock/scatter_detection_modelling.cxx:
```diff
@@ -75,15 +75,28 @@ find_detectors(unsigned& det_num_A, unsigned& det_num_B, const Bin& bin) const
   det_num_B =
     this->find_in_detection_points_vector(detector_coord_B + 
                                           this->shift_detector_coordinates_to_origin);
+  //std::cout << detector_coord_A << "\t" << detector_coord_B << "\n";
```
remove

KrisThielemans commented:
There's still some gremlins here. In addition, there's too many white-space changes. I'd prefer that you undo them. This PR should be as small as possible. At the moment, I have to append `?w=1` to the URL in the review as otherwise I'm overwhelmed. Even then there are quite a few changes that have nothing to do with code.

We'll solve the formatting after merging the PRs.

VietAnhDao committed:
remove comments

VietAnhDao committed:
undo & remove comment

VietAnhDao committed:
remove comment

VietAnhDao committed:
fixes and correct formating

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -76,17 +77,23 @@ SSRB(const ProjDataInfo& in_proj_data_info,
+    ProjDataInfoCylindrical * out_proj_data_info_sptr =
     dynamic_cast<ProjDataInfoCylindrical * >
     (in_proj_data_info_sptr->clone());
+  if (in_proj_data_info_sptr== NULL && out_proj_data_info_sptr==NULL)
```
```suggestion
  if (!in_proj_data_info_sptr )
```
is more C++11 like. you'll get warnings otherwise for some compilers anyway. Also no need to test for `out` as it's a `clone`

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -76,17 +77,23 @@ SSRB(const ProjDataInfo& in_proj_data_info,
     (in_proj_data_info_sptr->clone());
+  if (in_proj_data_info_sptr== NULL && out_proj_data_info_sptr==NULL)
+    {
+      warning("proj_data_info is not of type ProjDataInfoCylindrical will now try ProjDataInfoBlocksOnCylindrical\n");
```
```suggestion
      info("SSRB: proj_data_info is not of type ProjDataInfoCylindrical, will now try ProjDataInfoBlocksOnCylindrical", 2);
```
it's not a warning... adding the 2 should disable the message when using decreasing verbosity

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -76,17 +77,23 @@ SSRB(const ProjDataInfo& in_proj_data_info,
+      warning("proj_data_info is not of type ProjDataInfoCylindrical will now try ProjDataInfoBlocksOnCylindrical\n");
+      const ProjDataInfoBlocksOnCylindrical * const in_proj_data_info_sptr =
+    dynamic_cast<ProjDataInfoBlocksOnCylindrical const * >
+    (&in_proj_data_info);
```
now need to have a check
```
if (!in_proj_data_info_sptr)
  error("SSRB: blabla");

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -76,17 +77,23 @@ SSRB(const ProjDataInfo& in_proj_data_info,
+      const ProjDataInfoBlocksOnCylindrical * const in_proj_data_info_sptr =
+    dynamic_cast<ProjDataInfoBlocksOnCylindrical const * >
+    (&in_proj_data_info);
+    ProjDataInfoBlocksOnCylindrical * out_proj_data_info_sptr =
+    dynamic_cast<ProjDataInfoBlocksOnCylindrical * >
+    (in_proj_data_info_sptr->clone());
```
we now have a double clone and a memory leak. More worryingly, the `in_proj_data_info_sptr` and `out_proj_data_info_sptr` variables are local in scope, so they cannot be used outside. Is this code correct?

VietAnhDao replied:
yes since it's local it will not work for BoC data. What about?

```
in_proj_data_info_sptr.reset(dynamic_cast<ProjDataInfoBlocksOnCylindrical*>(&in_proj_data_info))
```

similarly with ```out_proj_data_info_sptr```

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -139,37 +150,49 @@ SSRB(const ProjDataInfo& in_proj_data_info,
-                 1))
-              warning("SSRB: in_proj_data_info with non-identical axial compression for all segments.\n"
-		      "That's ok, but results might not be what you expect.\n");
+        float min_m =1.E37F;
+        float max_m =-1.E37F;
+        for (int in_segment_num = in_min_segment_num; 
+            in_segment_num <= in_max_segment_num;
+            ++in_segment_num)
+          {
+                  if (in_axial_compression !=
+                      (in_proj_data_info_sptr->get_max_ring_difference(in_segment_num) -
+                      in_proj_data_info_sptr->get_min_ring_difference(in_segment_num) +
+                      1))
+                    warning("SSRB: in_proj_data_info with non-identical axial compression for all segments.\n"
+                "That's ok, but results might not be what you expect.\n");
```
still whitespace diffs

KrisThielemans replied:
and below

KrisThielemans commented on src/buildblock/SSRB.cxx:
```diff
@@ -208,18 +231,18 @@ SSRB(ProjData& out_proj_data,
   {
-    error("SSRB works only on segments with proj_data_info of "
-      "type ProjDataInfoCylindrical\n");
+    warning("proj_data_info is not of type ProjDataInfoCylindrical will now try ProjDataInfoBlocksOnCylindrical\n");
```
same problems as above I guess

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx:
```diff
@@ -676,15 +678,19 @@ calculate_proj_matrix_elems_for_one_bin(
-  const float costheta = 1/sqrt(1+square(tantheta));
+
+  float costheta = 1/sqrt(1+square(tantheta));
+  if (costheta < 0.0001) {
+    costheta = 0.0001;
+  }
```
can't see that we can ever have this (an LOR parallel to the scanner axis), but ok. do we need to threshold? Maybe add a comment why you do this

VietAnhDao replied:
printing out tantheta for BoC it produces Inf value. This is due to ```get_tantheta(const Bin& bin)``` from ProjDataInfoBlocksOnCylindrical.inl returning:

``` return delta_z/(2*sqrt(square(R)-square(get_s(bin))));```

The pesky radius returns again which causes problems when R = get_s(bin) which gives inf. A solution is to threshold the costheta and it seems to work.

KrisThielemans replied:
ok, but that seems to imply that your number of tang pos is too large. No point reconstructing a voxel that it as the "radius" (ok, more so in block-cylindrical of course) . In any case, it seems `tantheta` could just as well get into trouble with sqrt of a negative number then.

We should have a check on that upfront maybe, although that isn't so trivial as the projector probably doesn't (or shouldn't) know about R. a job for `ProjDataInfoBlocksOnCylindrical`? (e.g. in its `set_params` do a check on `get_s` for the outer bins). I'd rather do it there than in the projector. 

Because of the sqrt, I think the suggested work-around isn't sufficient.

VietAnhDao replied:
hmm an alternative is to generate an array of distance from origin to perimeter of scanner that varies for angles theta ( call this R(theta) ? ) when "BlocksOnCylindrical" object is created. Then we can use R(theta) as substitute for R throughout the "BlocksOnCylindrical" without involving the projector?

KrisThielemans replied:
can you put in a link to the `get_tantheta` code? Which `R` is it using anyway? Maybe that's the one that needs to be rewritten as you say.

VietAnhDao replied:
get_tantheta:

https://github.com/pkhateri/STIR/blob/b0e4f4a2959503b6d9ba0cd06d071248fa3a9adc/src/include/stir/ProjDataInfoBlocksOnCylindrical.inl#L101-L112

The R value is the inner "ring_radius", smallest distance from origin to perimeter initialised here:

https://github.com/pkhateri/STIR/blob/b0e4f4a2959503b6d9ba0cd06d071248fa3a9adc/src/buildblock/ProjDataInfoBlocksOnCylindrical.cxx#L71-L72

I believe all values from ```R=get_ring_radius(bin.view_num())``` returns the same value since it's not pre-calculated and assumes cylindrical geometry. Maybe the calculating vector of ```ring_radius``` is a good place to start.

KrisThielemans replied:
Here is what we think for `get_tantheta()` independent of `R`
```
CartesianCoordinate3D<float> _p1;
CartesianCoordinate3D<float> _p2;
find_cartesian_coordinates_of_detection(_p1, _p2, bin);
CartesianCoordinate3D<float> p2_minus_p1 = _p2 - _p1;
return p2_minus_p1.z() / (sqrt(square(p2_minus_p1.x())+square(p2_minus_p1.y());
```

VietAnhDao replied:
hmm it's really that simple... do you want me to commit this change?

KrisThielemans replied:
ok for me.

@danieldeidda, do we have a test on `get_tantheta` in `test_proj_data_info` already? (easiest one: compare the cylindrical and blocks give the same result for the identical scanner).

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -25,6 +25,10 @@
   \author Charalampos Tsoumpas
   \author Kris Thielemans
 */
+
+//#include <plog/Log.h> // Step1: include the headers
+//#include "plog/Initializers/RollingFileInitializer.h"
+
```
delete

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -120,11 +125,11 @@ process_data()
             bin_counter +=
                     this->proj_data_info_cyl_noarc_cor_sptr->get_num_axial_poss(vs_num.segment_num()) *
                     this->proj_data_info_cyl_noarc_cor_sptr->get_num_tangential_poss();
-            /* ////////////////// SCATTER ESTIMATION TIME ////////////////*/
```
reinstate

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -137,7 +142,6 @@ process_data()
                 previous_bin_count = bin_counter ;
                 wall_clock_timer.start();
             }
-            /* ////////////////// end SCATTER ESTIMATION TIME //////////////// */
```
reinstate

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -120,11 +125,11 @@ process_data()
              vs_num.view_num() <= this->proj_data_info_cyl_noarc_cor_sptr->get_max_view_num();
              ++vs_num.view_num())
         {
+            
```
delete

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -313,6 +318,7 @@ Succeeded
 ScatterSimulation::
 set_up()
 {
+
```
delete

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -401,25 +421,24 @@ void
-  const VoxelsOnCartesianGrid<float> & image = dynamic_cast<VoxelsOnCartesianGrid<float> const& >(_image);
-  const float z_to_middle =
-    (image.get_max_index() + image.get_min_index())*image.get_voxel_size().z()/2.F;
+   const VoxelsOnCartesianGrid<float> & image = dynamic_cast<VoxelsOnCartesianGrid<float> const& >(_image);
+   const float z_to_middle =
+      (image.get_max_index() + image.get_min_index())*image.get_voxel_size().z()/2.F;
```
white space diffs

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -401,25 +421,24 @@ void
     (scanner.get_num_rings()-1) * scanner.get_ring_spacing()/2;
 #endif
-  const VoxelsOnCartesianGrid<float> & act_image =
+   const VoxelsOnCartesianGrid<float> & act_image =
```
white space

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -401,25 +421,24 @@ void
+   const VoxelsOnCartesianGrid<float> & act_image =
     dynamic_cast<VoxelsOnCartesianGrid<float> const& >(*this->activity_image_sptr);
-  const float z_to_middle_standard =
+   const float z_to_middle_standard =
```
white space

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -401,25 +421,24 @@ void
     (act_image.get_max_index() + act_image.get_min_index())*act_image.get_voxel_size().z()/2.F;
-
-  if (abs(z_to_middle - z_to_middle_standard) > .1)
+    if (abs(z_to_middle - z_to_middle_standard) > .1)
```
white space

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -401,25 +421,24 @@ void
                         "This would cause a shift of %2%mm w.r.t. the activity image.\n"
                         "(see https://github.com/UCL/STIR/issues/495.")
-          % name % (z_to_middle - z_to_middle_standard));
+           % name % (z_to_middle - z_to_middle_standard));
```
white space

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -783,13 +806,13 @@ ScatterSimulation::downsample_scanner(int new_num_rings, int new_num_dets)
 	  {
 	    const float total_axial_length = proj_data_info_cyl_noarc_cor_sptr->get_scanner_sptr()->get_num_rings() *
 	      proj_data_info_cyl_noarc_cor_sptr->get_scanner_sptr()->get_ring_spacing();
-
```
white space

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -818,13 +841,11 @@ ScatterSimulation::downsample_scanner(int new_num_rings, int new_num_dets)
     // in ScatterEstimation. Otherwise use the max possible.
     int delta_ring = proj_data_info_cyl_noarc_cor_sptr->get_num_segments() == 1 ?  0 :
             new_scanner_sptr->get_num_rings()-1;
-
```
white space

KrisThielemans commented on src/scatter_buildblock/scatter_detection_modelling.cxx:
```diff
@@ -75,9 +75,21 @@ find_detectors(unsigned& det_num_A, unsigned& det_num_B, const Bin& bin) const
+        find_cartesian_coordinates_of_detection(
+                                                detector_coord_A,detector_coord_B,bin);
+    }else{
+      error("wrong type for scatter simulation");
```
```suggestion
      error("wrong type of projection data for scatter simulation");
```

KrisThielemans commented on src/scatter_buildblock/scatter_detection_modelling.cxx:
```diff
@@ -153,7 +165,7 @@ detection_efficiency_no_scatter(const unsigned det_num_A,
         error("ScatterSimulation::find_detectors: need to call set_up() first");
 #endif
-
+  
```
white space

VietAnhDao committed:
undo SSRB changes, will be a seperate PR.

VietAnhDao committed:
minor fix

VietAnhDao commented:
Completely reverted SSRB code. A proper implementation for SSRB will be given in separate PR.

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx:
```diff
@@ -761,9 +767,15 @@ calculate_proj_matrix_elems_for_one_bin(
+  float fovrad_in_mm = 
     min((min(max_index.x(), -min_index.x()))*voxel_size.x(),
         (min(max_index.y(), -min_index.y()))*voxel_size.y()); 
+        if (proj_data_info_ptr->get_scanner_ptr()->get_scanner_geometry() == "BlocksOnCylindrical")
+  {
+    fovrad_in_mm =
+        min((min(max_index.x(), -min_index.x()) - 5.f) * voxel_size.x(),
+            (min(max_index.y(), -min_index.y()) - 5.f) * voxel_size.y());
+  }
```
can you document why you need this?

VietAnhDao replied:
reducing the ```fovrad_in_mm``` prevent memory leak. Due to radius problem in "BlocksOnCylindrical" it might perform ray tracing where voxels doesn't exist. This might be the source of memory leak. I don't know too much about C++ technicality behind this but I know that if I perform reconstruction without ```-5.f``` or use ```+5.f``` then memory leak will occur. This I have tested.

KrisThielemans commented on src/scatter_buildblock/ScatterSimulation.cxx:
```diff
@@ -378,11 +380,20 @@ set_up()
+        if(dynamic_cast<ProjDataInfoBlocksOnCylindricalNoArcCorr*> (proj_data_info_cyl_noarc_cor_sptr.get())){
+            // BlocksOnCylindrical geometry is already align no shifted needed.
+            this->shift_detector_coordinates_to_origin =
+                CartesianCoordinate3D<float>(0,0,0);
```
In @danieldeidda's change, we shifted the "blocks" geometry to coincide with the cylindrical one. (i.e. 0 in first ring), so we will need to replace this code with the same as the lines above

KrisThielemans commented:
will need to add some code to `test_ScatterSimulation` to run the test with a block scanner. Likely needs adding code for the downsampling of such a scanner.

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx:
```diff
@@ -676,15 +678,19 @@ calculate_proj_matrix_elems_for_one_bin(
-  const float costheta = 1/sqrt(1+square(tantheta));
+
+  float costheta = 1/sqrt(1+square(tantheta));
+  if (costheta < 0.0001) {
+    costheta = 0.0001;
+  }
```
```suggestion
```

KrisThielemans replied:
@VietAnhDao please commit the above. I cannot push to your branch.

KrisThielemans commented on src/include/stir/ProjDataInfoBlocksOnCylindricalNoArcCorr.inl:
```diff
@@ -87,11 +87,65 @@ float
+  {
+    x = p0_minus_p1.x() - (p0_minus_p1_dot_p2_minus_p1/p2_minus_p1_magitude)*p2_minus_p1.x();
+    y = p0_minus_p1.y() - (p0_minus_p1_dot_p2_minus_p1/p2_minus_p1_magitude)*p2_minus_p1.y();
+  }
```
```suggestion
  }
  else
  {
     error("get_s(): 2 detection points are too close to each other. This indicates an internal error.")
  }
```

VietAnhDao committed:
Update src/include/stir/ProjDataInfoBlocksOnCylindricalNoArcCorr.inl

Co-authored-by: Kris Thielemans <KrisThielemans@users.noreply.github.com>

VietAnhDao committed:
Update src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx

Co-authored-by: Kris Thielemans <KrisThielemans@users.noreply.github.com>

KrisThielemans commented:
Thanks! Sadly, we cannot merge as there is a conflict in `src/buildblock/GeometryBlocksOnCylindrical.cxx`. `git` cannot resolve them most likely due to white-space changes (which is one reason I don't like them :-)).

In particular, `start_z` has been changed by @danieldeidda.

Can you merge `pkhateri/master` on yours and sort out the conflict? sorry.

VietAnhDao committed:
chage get_tantheta()

VietAnhDao committed:
semicolon

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx:
```diff
@@ -668,6 +669,7 @@ calculate_proj_matrix_elems_for_one_bin(
 
       phi = proj_data_info_noarccor.get_phi(bin);
       s_in_mm = proj_data_info_noarccor.get_s(bin);
+      tantheta = proj_data_info_noarccor.get_tantheta(bin);
```
why is it repeated here? You've already set it above. (in fact, `s_in_mm`  is also set above)

VietAnhDao replied:
it was set before I downcast the projection data so it would be accessible outside if statement and repeated inside the if statement to get correct value from each geometry of projection data.

KrisThielemans replied:
`get_tantheta` is `virtual`, so downcasting or not makes no difference: it will call the same function.

VietAnhDao replied:
didn't know this. What would happen if there is multiple ```get_tantheta()``` for each geometry?

KrisThielemans replied:
The pointer points to one particular object. Calling one of its virtual functions will call the one from its type, even if the pointer is from a base type. "run-time polymorphism" :-)

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx:
```diff
@@ -676,15 +678,16 @@ calculate_proj_matrix_elems_for_one_bin(
   
-  const float tantheta = proj_data_info_ptr->get_tantheta(bin);
-  const float costheta = 1/sqrt(1+square(tantheta));
+
+  float costheta = 1/sqrt(1+square(tantheta));
```
in fact, these lines can now be reinstated, and all the `tantheta` lines above removed, which will lead to a smaller change.

VietAnhDao replied:
Yeah sorry, I thought I made the changes to fix the white space problem so I marked as resolve. Not too sure if it's me or you who should be marking them as resolved, if it's not resolved just mark them as unresolved again.

As for the git conflict, do I do ```git merge master```? My plan is probably add Daniels changes then delete the "X" orientation code after.

KrisThielemans reviewed (changes requested):
I found 2 more things. sorry.

I also noticed you marked several of my requests to reinstate white-space/comment to the original lines as "resolved", but they are still there. Not sure why you did that. However, this is a minor thing, so feel free to ignore.

KrisThielemans commented:
> As for the git conflict, do I do git merge master? My plan is probably add Daniels changes then delete the "X" orientation code after.

what you should normally do is (assuming that you called  the remote pointing to Parisa's fork `pkatheri`)
```
git fetch pkhateri
git checkout BlockSTIR # just to make sure
git merge pkhateri/master
# resolve conflict by editing src/buildblock/GeometryBlocksOnCylindrical.cxx
git add src/buildblock/GeometryBlocksOnCylindrical.cxx
git commit
```
This will pull in Daniel's new test but also the conflicting edits, which is why you need to resolve them. (i.e. no need to remove the "X" configuration yet again)
git add

KrisThielemans commented:
> I thought I made the changes to fix the white space problem so I marked as resolve. Not too sure if it's me or you who should be marking them as resolved, if it's not resolved just mark them as unresolved again.

Looks like I cannot "unresolve". Have a look at the diff in `ScatterEstimation.cxx`.

VietAnhDao committed:
merged master to BlocksSTIR

KrisThielemans commented on src/buildblock/GeometryBlocksOnCylindrical.cxx:
```diff
@@ -156,11 +165,77 @@ build_crystal_maps()
 		see start_x*/
 
 		//calculate start_point to build the map.
+<<<<<<< HEAD
```
well. no. You've now committed the file that contains the conflict, i.e. it has both versions (your code here, and the original `pkhateri/master` code below). You need to manually edit the file.

VietAnhDao replied:
ok figured out what happened. VS code trying to be helpful shows the conflict as an overlay and some options to resolve the conflict. Unfortunately, unless I select an option it will keep both changes which is what it did :) .

VietAnhDao committed:
remove confict

KrisThielemans commented:
great. thanks. Can you sort out the `tantheta` comments in `src/recon_buildblock/ProjMatrixByBinUsingRayTracing.cxx`. Possibly the white-space/comment stuff in the scatter code. Then ready to go!

KrisThielemans commented:
@danieldeidda @emikhay this is nearly ready. Any objections from anyone to squash-merge (as opposed to merge) when done? The disadvantage is that it gets you into trouble if you've merged this PR somewhere else already.  The advantage however is that it removes a lot of the back and forth commits, so I'd prefer that.

If no problem, @VietAnhDao it'd help if you draft a commit message for the squash-merge, i.e. a few bullets with the actual changes that survived.

VietAnhDao committed:
Summary of chages:

- s coodrinate is derived from vectors.
- tantheta is derived from vectors.
- scatter simulation accept BlocksOnCylindrical.

VietAnhDao commented:
done :) . I hope I've removed all the white space.

VietAnhDao committed:
Summary of chages:

- s coodrinate is derived from vectors.
- tantheta is derived from vectors.
- scatter simulation accept BlocksOnCylindrical.
- remove X orientation.

VietAnhDao commented:
I have now :)

KrisThielemans commented:
Summary of the changes:
- get_s() and get_tantheta() is derived from detection coordinates for the BlocksOnCylindrical case
- scatter simulation accept BlocksOnCylindrical.

KrisThielemans commented:
Now I find that I cannot merge this...

@pkhateri could you squash-merge with the above commit message?

KrisThielemans commented:
@pkhateri or give me write permission to your `master` branch.

KrisThielemans commented:
I've squash-merged this on the `git` command line and pushed to (this) `master`. So this PR should be closed (but I cannot).

Thanks @VietAnhDao !

KrisThielemans commented:
yes please

VietAnhDao commented:
Pull request accepted.
